compiler/qsc_data_structures/src/namespaces.rs

```diff
@@ -0,0 +1,304 @@
+    fn default() -> Self {
+        let mut tree = Self {
+            assigner: 0,
+            tree: Rc::new(RefCell::new(NamespaceTreeNode {
+                children: FxHashMap::default(),
+                id: NamespaceId::new(0),
+            })),
+            memo: RefCell::new(FxHashMap::default()),
+        };
+        // insert the prelude namespaces using the `NamespaceTreeRoot` API
+        for ns in &PRELUDE {
+            let iter = ns.iter().map(|s| Rc::from(*s)).peekable();
+            tree.insert_or_find_namespace(iter);
+        }
+        tree
+    }
+}
+
+/// A node in the namespace tree. Each node has a unique ID and a map of children.
+/// Supports interior mutability of children for inserting new nodes.
+#[derive(Debug, Clone)]
+pub struct NamespaceTreeNode {
+    pub children: FxHashMap<Rc<str>, NamespaceTreeCell>,
+    pub id: NamespaceId,
+}
+impl NamespaceTreeNode {
+    /// Create a new namespace tree node with the given ID and children. The `id` should come from the `NamespaceTreeRoot` assigner.
+    #[must_use]
+    fn new(id: NamespaceId, children: FxHashMap<Rc<str>, NamespaceTreeCell>) -> Self {
+        Self { children, id }
+    }
+
+    /// Get a reference to the children of the namespace tree node.
+    #[must_use]
+    pub fn children(&self) -> &FxHashMap<Rc<str>, NamespaceTreeCell> {
+        &self.children
+    }
+
+    /// See [`FxHashMap::get`] for more information.
+    fn get(&self, component: &Rc<str>) -> Option<NamespaceTreeCell> {
+        self.children.get(component).cloned()
+    }
+
+    /// Get the ID of this namespace tree node.
+    #[must_use]
+    pub fn id(&self) -> NamespaceId {
+        self.id
+    }
+
+    /// Check if this namespace tree node contains a given namespace as a child.
+    #[must_use]
+    pub fn contains<'a>(&self, ns: impl IntoIterator<Item = &'a str>) -> bool {
+        self.get_namespace_id(ns).is_some()
+    }
+
+    /// Finds the ID of a namespace given its string name. This function is generally more efficient
+    /// than [`NamespaceTreeNode::find_namespace_by_id`], as it utilizes the prefix tree structure to
+    /// find the ID in `O(n)` time, where `n` is the number of components in the namespace name.
+    pub fn get_namespace_id<'a>(
+        &self,
+        ns: impl IntoIterator<Item = &'a str>,
+    ) -> Option<NamespaceId> {
+        let mut buf: Option<NamespaceTreeCell> = None;
+        for component in ns {
+            if let Some(next_ns) = match buf {
+                None => self.get(&Rc::from(component)),
+                Some(buf) => buf.borrow().get(&Rc::from(component)),
+            } {
+                buf = Some(next_ns);
+            } else {
+                return None;
+            }
+        }
+        Some(buf.map_or_else(|| self.id, |x| x.borrow().id))
+    }
+
+    /// Inserts a new namespace into the tree, if it does not yet exist.
+    /// Returns the ID of the namespace.
+    pub fn insert_or_find_namespace<I>(
```

Comment:
It might be worth noting in the doc string that this function only returns `None` when given an empty `iter`.